docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,833 @@
+all the race conditions with the domain iterator also apply here – i.e. the pointer loaded
+into the domain iterator, whose time range the unary iterator relies on to calculate the
+domain slices, may be incorrect or outdated.
+
+In step 2, this requires that the underlying file's data in the specified location by
+the offset and length found in step one has not changed – this is a possible
+race condition with Garbage Collection.
+
+In step 3, this involves moving the domain iterator, so all the race condition in the
+domain level also apply here.
+
+### 4.3.4 Deleting
+
+Deletion at the Unary level is a direct forward of deletion from the Domain level. The
+only added logic is the opening of an absolute gate on the time range being deleted –
+this will invalidate all deletions on ranges where a writer may write as well as all
+deletions whose time range intersect with another deletion's time range. Once the gate is
+created, since it has absolute authority, it will remain in control for the entirety of the
+delete duration.
+
+### 4.3.5 Garbage Collection
+
+Garbage collection at the Unary level is a direct forward of Garbage Collection from the Domain level.
+
+## 4.4 Cesium level race conditions
+
+At the Cesium level, multiple unaryDBs are stored to represent multiple channels. Therefore,
+reads and writes at this level culminate in **frames** (series from different channels).
+
+At the Cesium level, each entity may control multiple channels – i.e. multiple Unary or
+Virtual databases. Fortunately, the execution of these entities on different channels do
+not interfere with each other: a writer on one unary channel will not cause race conditions
+on another channel. Similarly, Iterator at the Cesium level is simply a stream wrapper
+that forwards the commands it receives to its unary-level internal iterators.
+
+Out of the four entities, the only entity that requires particular attention at the Cesium
+level is the deleter, as we must not delete index unaryDBs that contain time stamp data
+for other data channels. To this end, we first check under an absolute controller region
+that the time range is neither controlled by a writer (in which case we would report that
+there is data in that time range) nor containing any time stamp data. If there are no data,
+then we call the unaryDB's `delete` method to begin deleting data. Note that there is a
+race condition here: right after our `HasDataFor`'s gate is released,
+there could be a writer that writes some data and closes its gate right before `delete`'s
+gate is acquired, causing us to delete index data for other channels. However, we do not
+anticipate seeing this case often, as very rarely would one delete an index channel but
+keep some data channels indexed by that index channel.
+
+## 4.5 Conclusion
+
+Here is a list of race conditions revealed from this RFC:
+
+- Domain level:
+  - file_controller: acquireWriter's descriptor limit check may not extend into the following
```

Comment:
"file_controller", "acquireWriter" should be code. Also, refer to them as databases instead of levels. Finally, remove capitalization from things like "Garbage Collection".